File: service/app/sources/hud.py
"""HUD Fair Market Rents (free, unlimited). ZIP -> county (USPS crosswalk) ->
county FMR record (Small Area FMRs per ZIP). Ported from the app's TS version."""
from __future__ import annotations

from typing import Optional

from .. import cache
from ..config import settings
from ..http import client

BASE = "https://www.huduser.gov/hudapi/public"
FMR_YEARS = [2026, 2025, 2024]
BED_KEY = {0: "Efficiency", 1: "One-Bedroom", 2: "Two-Bedroom", 3: "Three-Bedroom", 4: "Four-Bedroom"}


async def _get(path: str) -> Optional[dict]:
    if not settings.hud_api_token:
        return None
    try:
        async with client(headers={"authorization": f"Bearer {settings.hud_api_token}"}) as c:
            r = await c.get(f"{BASE}{path}")
            if r.status_code != 200:
                return None
            return r.json()
    except Exception:
        return None


async def _zip_to_county(zipcode: str) -> Optional[str]:
    data = await _get(f"/usps?type=2&query={zipcode}")
    results = (data or {}).get("data", {}).get("results", [])
    if not results:
        return None
    results.sort(key=lambda r: r.get("res_ratio", 0), reverse=True)
    return results[0].get("geoid")


async def fmr_by_zip(zipcode: str) -> Optional[dict]:
    cached = cache.get("hud_fmr", zipcode)
    if cached is not None:
        return cached or None
    county = await _zip_to_county(zipcode)
    if not county:
        return None
    data = None
    for year in FMR_YEARS:
        j = await _get(f"/fmr/data/{county}99999?year={year}")
        if j and j.get("data"):
            data = j["data"]
            break
    if not data:
        return None
    bd = data.get("basicdata")
    arr = bd if isinstance(bd, list) else [bd]
    entry = (
        next((b for b in arr if str(b.get("zip_code")) == zipcode), None)
        or next((b for b in arr if "msa" in str(b.get("zip_code")).lower()), None)
        or (arr[0] if arr else {})
    )
    rent_by_bed = {}
    for beds, key in BED_KEY.items():
        v = entry.get(key)
        if v:
            rent_by_bed[beds] = int(v)
    out = {
        "year": data.get("year"),
        "area_name": data.get("area_name") or data.get("county_name"),
        "small_area": str(data.get("smallarea_status")) == "1",
        "rent_by_bed": rent_by_bed,
    }
    cache.put("hud_fmr", zipcode, out)
    return out


def fmr_for_beds(rent_by_bed: dict, beds: int) -> Optional[int]:
    if not rent_by_bed:
        return None
    b = max(0, min(4, beds))
    return rent_by_bed.get(b) or rent_by_bed.get(str(b)) or next(iter(rent_by_bed.values()), None)

File: src/lib/server/hud.ts
// HUD Fair Market Rents (FMR) API — free U.S. government data, unlimited.
// Needs a free access token from https://www.huduser.gov/portal/dataset/fmr-api.html
// Key resolution: header `x-hud-key` → env HUD_API_TOKEN.
//
// FMR ≈ the 40th-percentile gross rent HUD publishes for Section 8 — a solid,
// conservative market-rent BASELINE (often a bit under nice-unit market rent).
//
// Lookup flow (the direct /data/{zip} endpoint is unreliable):
//   1. ZIP → county FIPS via the USPS ZIP→county crosswalk.
//   2. county FIPS → FMR via /fmr/data/{fips}99999 (returns Small-Area FMRs
//      per ZIP for metro areas, plus an MSA-level fallback row).

const BASE = "https://www.huduser.gov/hudapi/public";
const FMR_YEARS = [2026, 2025, 2024]; // try newest published first

export class HudError extends Error {
  status: number;
  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

export function resolveHudToken(req: Request): string | undefined {
  return req.headers.get("x-hud-key") || process.env.HUD_API_TOKEN || undefined;
}

const BED_KEY: Record<number, string> = {
  0: "Efficiency",
  1: "One-Bedroom",
  2: "Two-Bedroom",
  3: "Three-Bedroom",
  4: "Four-Bedroom",
};

export interface FmrResult {
  year?: string | number;
  areaName?: string;
  smallArea: boolean;
  rentByBed: Record<number, number>;
}

async function hudGet(path: string, token: string): Promise<{ ok: boolean; status: number; json: unknown }> {
  const res = await fetch(`${BASE}${path}`, {
    headers: { Authorization: `Bearer ${token}`, accept: "application/json" },
  });
  let json: unknown = null;
  try {
    json = await res.json();
  } catch {
    /* non-JSON */
  }
  return { ok: res.ok, status: res.status, json };
}

async function zipToCounty(zip: string, token: string): Promise<string | undefined> {
  const { ok, status, json } = await hudGet(`/usps?type=2&query=${zip}`, token);
  if (!ok) {
    if (status === 401) throw new HudError("HUD token rejected (401). Re-check the token in Settings.", 401);
    return undefined;
  }
  const results = (json as { data?: { results?: { geoid?: string; res_ratio?: number }[] } })?.data?.results ?? [];
  if (!results.length) return undefined;
  results.sort((a, b) => (b.res_ratio ?? 0) - (a.res_ratio ?? 0));
  return results[0].geoid;
}

export async function fmrByZip(zip: string, token: string): Promise<FmrResult> {
  const county = await zipToCounty(zip, token);
  if (!county) throw new HudError(`No HUD county mapping found for ZIP ${zip}.`, 404);

  let data: Record<string, unknown> | undefined;
  for (const year of FMR_YEARS) {
    const { ok, json } = await hudGet(`/fmr/data/${county}99999?year=${year}`, token);
    if (ok && (json as { data?: unknown })?.data) {
      data = (json as { data: Record<string, unknown> }).data;
      break;
    }
  }
  if (!data) throw new HudError("HUD returned no FMR data for that area.", 502);

  const bdRaw = data.basicdata;
  const arr = Array.isArray(bdRaw) ? (bdRaw as Record<string, unknown>[]) : [bdRaw as Record<string, unknown>];
  const entry =
    arr.find((b) => String(b?.zip_code) === zip) ??
    arr.find((b) => String(b?.zip_code).toLowerCase().includes("msa")) ??
    arr[0] ??
    {};

  const rentByBed: Record<number, number> = {};
  for (const [beds, key] of Object.entries(BED_KEY)) {
    const v = Number(entry[key]);
    if (!Number.isNaN(v) && v > 0) rentByBed[Number(beds)] = v;
  }

  return {
    year: data.year as string,
    areaName: (data.area_name as string) || (data.county_name as string) || undefined,
    smallArea: String(data.smallarea_status) === "1",
    rentByBed,
  };
}

/** Pick the FMR for a bedroom count, clamping to the available range (4 = 4+). */
export function fmrForBeds(rentByBed: Record<number, number>, beds: number): number | undefined {
  const b = Math.max(0, Math.min(4, beds));
  return rentByBed[b] ?? rentByBed[Math.min(4, b)] ?? Object.values(rentByBed)[0];
}
